### ISSUE TITLE
Encrypted ClientHello

### ISSUE THREAD

**Issue by hl2guide** [AdguardTeam/AdGuardHome#1681]
### Problem Description
Please consider adding support for **Encrypted SNI** in Adguard Home (if applicable).

I'm no network engineer so I don't know if it even applies, but it would be nice if it could be implemented.

CloudFlare's Reference: https://blog.cloudflare.com/encrypted-sni/
CloudFlare's Test Page: https://www.cloudflare.com/ssl/encrypted-sni/

**Comment by jkle112** [comment]
True!
Hope more security!
https://allinfa.com/dns-over-https-doh.html
https://allinfa.com/encrypted-server-name-indication-esni.html
https://www.cloudflare.com/ssl/encrypted-sni/

**Comment by RainmakerRaw** [comment]
For your own queries do you mean? Surely eSNI is already supported. I get _esni queries answered in my query log all the time, and [Cloudflare confirms](https://www.cloudflare.com/ssl/encrypted-sni/) my connections include encrypted SNI. 

Added to that, my ISP blocks torrent sites (court order, UK) but with eSNI enabled in Firefox, and AGH set to use a capable upstream (i.e. Cloudflare), I can load the sites no problems even without a VPN (because the ISP can't see what I'm connecting to, as both DNS and SNI are encrypted).

**Comment by ameshkov** [comment]
ESNI standard is far from being out of the draft state. FF and Cloudflare have implemented some old draft which is rather far from the current state of the spec.

Anyways, we can only implement this when ESNI is finalized and supported by Golang, and this is not the case yet.

**Comment by hbednar** [comment]
It looks like Cloudflare have now enabled it and i think it's on by default for new customers:
https://blog.cloudflare.com/announcing-encrypted-client-hello/

**Comment by ameshkov** [comment]
Not that simple, they seem to have disabled it for most of the users now, was causing some troubles.

Anyways, in golang there's no way to use ECH unless you're using a Cloudflare's fork of Go.

Generally it's possible to use it now, but I'd prefer to wait until ECH makes it to Go's `crypto/tls`.

**Comment by ameshkov** [comment]
Reopened just to show that we're planning to implement it eventually.

**Comment by vdias** [comment]
Any news on this topic?

**Comment by GalaxyVOID** [comment]
I am using ``tls://1dot1dot1dot1.cloudflare-dns.com`` as the main upstream DNS server and ``https://dns.cloudflare.com/dns-query`` for the fallback DNS server. I do get Encrypted SLI.

**Comment by ameshkov** [comment]
> Any news on this topic?

ECH is still not supported natively by `crypto/tls`. Switching all crypto to Cloudflare's fork is too much.

We probably could come up with a solution similar to what [I did in gocurl](https://github.com/ameshkov/gocurl/blob/master/internal/client/cfcrypto/cfcrypto.go) and implement it for upstreams.

> `https://dns.cloudflare.com/dns-query`

This domain does not indicate ECH support at the moment.

**Comment by donald2612** [comment]
I can stress this not enough, how important the progress is.

It's about federal-surveillance and such. Mostly I can disable all VPN-connections having that enabled in effect.

**Comment by lordraiden** [comment]
> > Any news on this topic?
> 
> ECH is still not supported natively by `crypto/tls`. Switching all crypto to Cloudflare's fork is too much.
> 
> We probably could come up with a solution similar to what [I did in gocurl](https://github.com/ameshkov/gocurl/blob/master/internal/client/cfcrypto/cfcrypto.go) and implement it for upstreams.
> 
> > `https://dns.cloudflare.com/dns-query`
> 
> This domain does not indicate ECH support at the moment.

So how will work the final implementation ?
Will we need to modify something in the browser or Adguard home will handle everything vía configuration?

As far as I know DNScrypt also solves this issue but it looks like is not working properly

https://github.com/AdguardTeam/AdGuardHome/issues/6897

I think this is an important thing to bring

**Comment by ezcompanynz** [comment]
I'm passing the SNI test on Cloudflare using DOH and Brave browser.
One thing to check is that you don't have DNSSEC enabled in Adguard as Cloudflare already does this (if you're using Cloudflare as your upstream provider) mine would fail before disabling this.

**Comment by gabriel-vanca** [comment]
Any update on this?

**Comment by xenstar** [comment]
@ameshkov ECH Widely adopted now. Please could you add ECH Support in Adguard Home. We have been waiting for many years. 

> The news are:
> 
> 1. It is now called ECH
> 2. It is still in the draft stage
> 3. It is still used by no one. CloudFlare even disabled it for now.

**Comment by xenstar** [comment]
> Anyways, we can only implement this when ESNI is finalized and supported by Golang, and this is not the case yet.

@ameshkov Yes, Go now supports ECH (Encrypted Client Hello) in its standard crypto/tls package. Client-side support landed in Go 1.23 (August 2024), and server-side support was added in Go 1.24 (February 2025).

**Comment by NielsNL12** [comment]
> > Anyways, we can only implement this when ESNI is finalized and supported by Golang, and this is not the case yet.
> 
> [@ameshkov](https://github.com/ameshkov) Yes, Go now supports ECH (Encrypted Client Hello) in its standard crypto/tls package. Client-side support landed in Go 1.23 (August 2024), and server-side support was added in Go 1.24 (February 2025).

It's also an official standard now.

https://www.rfc-editor.org/rfc/rfc9849.html

**Comment by voidblack** [comment]
+1 please implement ECH
